### ISSUE TITLE
Major Code Cleanup

### ISSUE THREAD

**Issue by Master-Foo** [Master-Foo/game#66]
This is important info, so everyone reply so I know you've grokked this fully.

OK, so I decided to stop being a hypocrite and clean up my messes. The overall structure of the program is mostly clean. So, now when someone submits code that isn't structurally sound I won't feel guilty for not leading by example.

As a result, the directory structure and MVC paradigm shouldn't change too much for the rest of the development process. So, now that I'm fairly certain that things are the way they should be, I'll explain where everything goes and why.

1. The root directory no longer has the executables.
    1. They have all been moved to `/demo`
    1. You can still run them, but you must be in the project root to do so for instance `./demo/build_workflow.py`
    1. The exception to this is the game itself.
        1. It is also in `./demo/demo.py`
        1. However, in the project root there is a `run` file. `./run` will execute the game.
1. The `/libs` folder is for 3rd party libraries
1. `/resources` has remained unchanged
1. `/requirements.txt` is for keeping track of python libraries. `pip` uses them to install dependencies.
1. `/run` is the game executable. (It is a bash script so @fishscene check to see if it works)
1. `/source` has many changes.
    1. `abstract` is for keeping classes that are **only ever inherited**, but never actually used directly in the game. Think of everything in this folder as like the primitive "lizard" part of our brain. It's mostly "instinctual" in nature. 
        1. `app` this is the core logic of the game. Things that every game should need and rarely be changed.
        1. `base_object` Almost everything will inherit from this.
        1. `chemistry` these are the abstract files for chemistry
        1. `entities` mostly unchanged, except I've cleaned up the MVC
        1. `location` a location is basically a system to determine where things are.
    1. `concrete` These are classes that are directly used in the game. Think of this as the higher order part of the brain as opposed to the "lizard brain" of `abstract`
        1. `entities` are mostly unchanged, except, I've cleaned up the MVC paradigm.
    1. `global_variables` globals are evil. Never use them. (unless you know the secret (if you need to be told the secret, it's "don't use globals, they are evil") )
    1. `library` these are for code that is useful in a general sense. If it isn't inherited, and isn't something that is generally "rendered" it probably goes in here.
        1. `action` this is for the rendering of sprites.
        1. `science` this holds code that is sciency.
            1. `chemistry` when you need to use anything chemistryish in the game. Most likely this is where you will find it.
            1. `geology` I broke out @jrslagle's rocks into here. Makes more sense.
            1. `math` nothing here yet, but if we ever need custom math functionality that doesn't already exist in another library, put it here.
        1. `solid_noise` this is the library we'll use for most of our procedural stuff.
        1. `ui` user interface stuff. Things like windows, keyboards, joysticks, cameras, etc.
    1. `systems` this is for in-game objects who's purpose is to hold other objects and give them the ability to communicate with other objects.
        1. `game` this is for less generic features of the game as opposed to `app` which is very generic.
        1. `location` these are the hierarchy of locations from the universe `Down` to the `Hectare` Used to determine were things are.
        1. `logistics` anything that is part of the "build workflow" will probably go in here.
        1. `power_grid` any electronics can connect to this and provide, store, or use energy.

Anyway, that is a general overview of where everything should go.

One thing I should note is that **the directory structure isn't consistent with inheritance structure** This is on purpose. **The directory structure is organized logically** So, the `plant` entities aren't put in `/source/concrete/entities/plant` because they inherit from `Plant`. They are put there to be grouped logically. It has nothing to do with actual inheritance. (Although, they do all inherit from `Plant`. It just so happen that "inheritance" is a convenient way to group something "logically") But don't confuse "logical" grouping with inheritance. The directory structure is just to keep packages in an orderly fashion.

A note on MVC.

Now that MVC has been cleaned up. I expect that to be the standard for now on. Look at some examples in the project. The best one to look at is Avatar. It's the most complicated Entity we have.

If someone wants to make a skel of an Entity so, we can just copy and paste a folder and change a couple of details in the files to make a new entity, go for it.

Also, `animation_config` has been moved into the `view` folder for each entity, because, logically, that is view related info.

**TL;DR** Don't TL;DR, this is important stuff. Mostly it means that the framework has matured to the point where I'm comfortable with it that I'm willing to take the time to explain everything. Whereas, before, much was still up for debate as far as structure. Now, it's pretty much the way it should be.

**Comment by Master-Foo** [comment]
Oh, yeah. Shouldn't have to say this, but with a change like this, it's probably best to do a pull before committing any code. (You should be doing a pull before every commit anyway because that's how you prevent commit conflicts)

**Comment by Master-Foo** [comment]
@jrslagle FYI: according to this [commit log](https://github.com/Master-Foo/game/commit/4f35516bac7550075baf2ceb5ae28ebd815236b0)

**361 changed files with 1,846 additions and 291 deletions. **

I did this in about 3ish hours. Don't take my word when I say that breaking files up won't slow you down. Here's proof. Just use the correct tools. Extrapolate that data out to the Elements and you have less than an hour of work to do everything that needs to be done in Elements.

**Comment by jrslagle** [comment]
@Master-Foo, yea I'm still not sure where to even start when it comes to splitting a file into 95 pieces. I'm still discovering new features all the time though.

Sounds like a plan. I picked through your new layout and it's all very logical. I've been meaning to find an MVC conventions reference (to head off some annoying questions), but most things I can think to add fit nicely into the structure you already have.

**Comment by Master-Foo** [comment]
There isn't any one "true" MVC convention. Every framework is going to have different ways of implementing MVC. All MVC is, is "break code up logically into different skill sets.". So, learning RoR, or Django's conventions could be useful, but even they differ from each other.

As for breaking the files up. It's just a matter of the tool you use. I use Sublime Text. It's like vim, but not annoying. The nice thing about it is you can do pretty much everything without a mouse. Using a mouse is 90% of what is slowing you down most likely. Then next 10% is probably just learning how to open files quickly on demand and doing mass search / replace.
